feos-dft/src/profile.rs

```diff
@@ -389,110 +383,71 @@ where
             });
 
         // calculate bond integrals
-        let mut exp_dfdrho = dfdrho.mapv(|x| (-x).exp());
+        let exp_dfdrho = dfdrho.mapv(|x| (-x).exp());
         let bonds = self
             .dft
             .bond_integrals(temperature, &exp_dfdrho, &self.convolver);
-        exp_dfdrho *= &bonds;
+        let mut rho_projected = &exp_dfdrho * bonds;
 
-        // Euler-Lagrange equation
-        res_rho
+        // multiply bulk density
+        rho_projected
             .outer_iter_mut()
-            .zip(exp_dfdrho.outer_iter())
             .zip(bulk_density.iter())
-            .zip(density.outer_iter())
-            .for_each(|(((mut res, df), &rho_b), rho)| {
-                res.assign(
-                    &(if log {
-                        rho.mapv(f64::ln) - rho_b.ln() - df.mapv(f64::ln)
-                    } else {
-                        &rho - rho_b * &df
-                    }),
-                );
+            .for_each(|(mut x, &rho_b)| {
+                x *= rho_b;
             });
 
+        // calculate residual
+        let mut res = if log {
+            rho_projected.mapv(f64::ln) - density.mapv(f64::ln)
+        } else {
+            &rho_projected - density
+        };
+
         // set residual to 0 where external potentials are overwhelming
-        res_rho
-            .iter_mut()
+        res.iter_mut()
             .zip(self.external_potential.iter())
             .for_each(|(r, &p)| {
```

Comment:
(not a recent change but) would this be more clear with `filter().for_each()` instead of the `if` case?

Reply:
Me gusta!